compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -1521,6 +1517,15 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
 
     #[instrument(skip(self), level = "debug")]
     fn report_fulfillment_error(&self, error: &FulfillmentError<'tcx>) -> ErrorGuaranteed {
+        let mut error = FulfillmentError {
+            obligation: error.obligation.clone(),
+            code: error.code.clone(),
+            root_obligation: error.root_obligation.clone(),
+        };
+        if self.apply_do_not_recommend(&mut error.obligation) {
+            error.code = FulfillmentErrorCode::Select(SelectionError::Unimplemented);
```

Comment:
I don't think this is correct `Select(Unimplemented)` is only meant for trait goals, and only if the goal is truly unimplemented (e.g. not ambiguity, not overflow, not a cycle).

I think we need to look at the `obligation` to re-derive the `FulfillmentErrorCode` -- perhaps look at how the `FulfillmentErrorCode` is created in the first place in the old solver?

Reply:
For the record, I haven't looked deep enough to see if this is possible to turn into a weird diagnostic (or an ICE), so you may need to do a bit of research yourself.

Reply:
> I think we need to look at the obligation to re-derive the `FulfillmentErrorCode` -- perhaps look at how the FulfillmentErrorCode is created in the first place in the old solver?

I'm not sure if that is easily possible with the old solver. As far as I understand these `FulfillmentErrorCode` are produced somewhat ad hoc in the code here that does the actual solving: 

https://github.com/rust-lang/rust/blob/f2e1a3a80ae54734e1a3d306f31c2caebb05de9b/compiler/rustc_trait_selection/src/traits/fulfill.rs#L745-L750

At this point we know the following things about the passed obligation:

* If it changes it does contain a unimplemented trait requirement somewhere in the stack as `#[do_not_recommend]` an only applied to trait impls (That's asserted at other places).
* I think we can only ever end up with changing anything here for `FulfillmentErrorCode::Project` or `FulfillmentErrorCode::Project`. In both cases it will change it to the a trait requirement not implemented obligation due to where `do_not_recommend` can be placed. 

Maybe it's a good idea to just explicitly check the `FulfillmentErrorCode` before if its `FulfillmentErrorCode::Select(Unimplementd)` or `FulfillmentErrorCode::Project` so that it always matches that assumptions? 


> For the record, I haven't looked deep enough to see if this is possible to turn into a weird diagnostic (or an ICE), so you may need to do a bit of research yourself.

At least all of the compiler test suite (via `./x.py test`) runs fine with this change. That said I will try to produce a few edge cases with the `#[do_not_recommend]` attribute added and add them as well to the tests.

Reply:
I've tried to think about potential edge cases in the last few days, but I did not came up with something interesting with the current set of restrictions.
We only walk the tree now for `FulfillmentErrorCode::Select(Unimplementd)` or `FulfillmentErrorCode::Project`. We only change anything if we hit a trait impl that is marked as `#[do_not_recommend]` at all. As these are enforced to be on impls I cannot see a different case than just having an Obligation that just says that a certain trait is not implemented.

@rustbot ready